spatial-lib/src/cmr/spatial/circle.clj

```diff
@@ -0,0 +1,106 @@
+(ns cmr.spatial.circle
+  (:require
+   [cmr.spatial.math :as math :refer :all]
+   [primitive-math]
+   [cmr.spatial.point :as p]
+   [cmr.spatial.derived :as d]
+   [cmr.spatial.geodetic-ring :as gr]
+   [cmr.spatial.validation :as sv]
+   [cmr.spatial.polygon :as poly]
+   [cmr.common.dev.record-pretty-printer :as record-pretty-printer])
+  (:import
+   cmr.spatial.geodetic_ring.GeodeticRing
+   cmr.spatial.polygon.Polygon
+   cmr.spatial.point.Point))
+
+(primitive-math/use-primitive-operators)
+
+;; Circle
+(defrecord Circle
+  [
+   ^Point center
+   ^double radius
+   ])
+
+(record-pretty-printer/enable-record-pretty-printing Circle)
+
+(defn circle
+  "Creates a new minimum bounding rectangle"
+  ([^Point center ^double radius]
+   (->Circle center radius))
+  ([^double lon ^double lat ^double radius]
+   (->Circle (p/point lon lat) radius)))
+
+(def MIN_RADIUS
+  "Minimum radius in meters"
+  10)
+
+(def MAX_RADIUS
+  "Maximum radius in meters"
+  1000000)
+
+(def ^:const ^double EARTH_RADIUS_APPROX
+  "Radius of the earth in meters for polygon approximation of the circle"
+  6378137.0)
+
+(defn- validate-center
+  "Validate the center of the circle, returns error if it is on the poles; otherwise returns nil"
+  [^Point center]
+  (let [{:keys [lon lat]} center]
+    (when (or (= 90.0 lat) (= -90.0 lat))
+      [(format "Circle center cannot be the north or south pole, but was [%s, %s]" lon lat)])))
+
+(defn validate-radius
+  "Validate the radius. Returns a list of errors when radius is invalid; otherwise returns nil"
+  [^double radius]
+  (when-not (<= (double MIN_RADIUS) radius (double MAX_RADIUS))
+    [(format "Circle radius must be between %s and %s, but was %s." MIN_RADIUS MAX_RADIUS radius)]))
+
+(defn covers-point?
+  "Returns true if the circle contains the given point"
+  [^Circle cir ^Point point]
+  (let [{:keys [^Point center ^double radius]} cir]
+    (>= radius (p/distance center point))))
+
+(defn circle->polygon
+  "Returns the polygon approximation of the circle with the given number of points.
+   Reference: https://github.com/gabzim/circle-to-polygon."
+  [^Circle cir ^long n]
+  (let [{:keys [^Point center ^double radius]} cir
+        lon1 (.lon_rad center)
+        lat1 (.lat_rad center)
+        r-rad (/ radius EARTH_RADIUS_APPROX)
+        points (persistent!
```

Comment:
This should be its own function with a docstring and whatnot - there is a lot of logic here

Reply:
The function is small already, most of the code is just calling this piece of code that does the math. It will be more unreadable to have it in its own function. The docstring has listed the reference already.

Reply:
Right, but this function has its own `let` - we usually want to try to keep such complex logic out of `let` blocks for readability/maintainability. I'll still approve, but I disagree with you here. The logic looks good, though :)